### DISCUSSION TITLE
Tutorial: Moving applications to external drive

### DISCUSSION THREAD

**Discussion by LXY113** [PlayCover/PlayCover#1712] [Show and tell]
Rough English translations provided through DeepL (with some modifications for readability from TheMoonThatRises).

This tutorial is written in Chinese, please use a web translator for other languages.
由于本人的Macbook Air只有256GB内存，所以准备把PlayCover软件移动到外接硬盘。
外接硬盘进行游戏需要较高的读写速度，经测试1000mb每秒的速度是与内置储存运行比较无感的。
下面开始教程：
1.由于我的游戏已经安装在了本地硬盘，所以本教程是移动方法的说明。打开「磁盘工具」软件，定位到你的移动硬盘，使分区格式为APFS。我们要将每个游戏放置在单独的APFS容器中，所以在将分区调整为APFS后，点击上方「宗卷」上面的「+」号，添加一个APFS容器，命名为游戏名（其实什么名字都可以），并完成创建。
2.以下操作中，你的用户名将以「abc」代替。首先来移动PlayCover本体，按照第一步的方法，创建PlayCover的APFS容器，我的就叫PlayCover。打开「终端」软件，使用「diskutil list」命令，查询外接硬盘的APFS容器编号。

> Since my Macbook Air only has 256GB of RAM, I'm going to move the PlayCover software to an external hard drive.
An external hard disk requires high read/write speeds to play games, and the tested speeds of 1000mb per second are comparable to running on internal storage.

> Here is the tutorial:
> 1. Since my game has been installed in the local hard disk, so this tutorial is to move the method of instructions. Open the “Disk Utility” software, locate your removable hard disk, and make the partition format APFS. we want to place each game in a separate APFS container, so after adjusting the partition to APFS, click on the “+” sign above the top of the “Volume” to add an APFS container, name it as the game name (actually any name will do), and finish creating it.
> 2. In the following operation, your username will be replaced by “abc”. First of all, let's move the PlayCover container, follow the first step to create the APFS container of PlayCover, mine is called PlayCover. open the “Terminal” software, use the “diskutil list” command to query the APFS container number of the external hard disk.

<img width="682" alt="截屏2024-10-10 08 00 43" src="https://github.com/user-attachments/assets/d7067226-7bf9-49eb-9130-c0b58b8995bf">

我们可以看到，外接硬盘为「disk5」，PlayCover容器为「disk5s2」。我们此时可以在桌面上看见PlayCover盘，我们打开「访达」在菜单栏的「前往」中选取「前往文件夹」，在地址栏键入地址「/Users/abc/Library/Containers」回车进入目录，找到PlayCover文件夹，打开后复制Data文件夹放入到桌面上的PlayCover盘，然后返回上一层即Containers文件夹内，右键PlayCover文件夹，选择「显示简介」，复制「名称与扩展名」下方的内容（比如PlayCover为「_io.playcover.PlayCover_」。使用命令「sudo umount -f /Volumes/PlayCover」卸载PlayCover容器，然后使用命令「sudo mount_apfs /dev/disk5s2 /Users/abc/Library/Containers/_io.playcover.PlayCover_」将创建的PlayCover的APFS容器装载到Containers的文件夹中（如果提示mount apfs: volume could not be mounted: Operation not permitted，请尝试关闭SIP）。然后启动PlayCover，如果系统提示PlayCover想访问外接硬盘的文件就说明移动成功，可在推出移动硬盘后将Containers文件夹中的PlayCover文件夹里面的Data删除来节省空间（也可能不提示，右键PlayCover里面的游戏选择在访达中查看，如果打开的是移动硬盘的路径说明成功）。
3.游戏的移动，例子是崩坏星穹铁道。具体方法和方法二类似，在「磁盘工具」软件中定位你的移动硬盘，在APFS分区中新建APFS容器，命名为「StarRail」（随便取名）。将Containers文件夹中的「崩坏：星穹铁道」中的Data文件夹复制到桌面上的StarRail容器中，使用「sudo umount -f /Volumes/StarRail」，然后「sudo mount_apfs /dev/disk5s3 /Users/abc/Library/Containers/_com.miHoYo.hkrpg_」（_com.miHoYo.hkrpg_就是Containers文件夹下崩坏：星穹铁道的文件夹显示简介中的名称与扩展名，disk5s3是我的StarRail容器编号，每个人不一样）。启动PlayCover中的「崩坏：星穹铁道」，如果系统提示游戏想访问外接硬盘的文件就说明移动成功，可在推出移动硬盘后将Containers文件夹中的「崩坏：星穹铁道」文件夹里面的Data删除来节省空间。

> We can see that the external hard disk is “disk5” and the PlayCover container is “disk5s2”. We can see the PlayCover disk on the desktop at this time, we open the “Access” in the menu bar of the “Go to” select “Go to folder”, in the address bar type in the address “/Users/abc/Library/Containers” Enter into the directory, to find the PlayCover folder, open the Data folder to copy and put it into the desktop PlayCover disk, then return to the previous level that is the Containers folder, right-click the PlayCover folder, and put it into the PlayCover folder on the desktop. Open the PlayCover folder, copy the Data folder and put it into the PlayCover disk on the desktop, then go back to the previous level, i.e. the Containers folder, right-click on the PlayCover folder, select “Show Profile”, and copy the content under “Name and Extension” (e.g. PlayCover is “io.playcover.PlayCover”). Unmount the PlayCover container using the command “sudo umount -f /Volumes/PlayCover” and then use the command “sudo mount_apfs /dev/disk5s2 /Users/abc/Library/Containers/io.playcover.PlayCover” to mount the created APFS container of PlayCover into the Containers folder (if prompted mount apfs: volume could not be mounted: Operation not permitted, please try to turn off SIP). Then start PlayCover, if the system prompts PlayCover wants to access the files of the external hard disk, it means the move is successful, you can delete the data inside the PlayCover folder in Containers folder after launching the removable hard disk to save the space (or you may not be prompted, right-click on the game inside PlayCover and select View in Access). If the path of the removable hard disk is opened, it means it is successful).

> 3. The movement of the game, the example is Honkai: Star Rail. The specific method is similar to method 2, locate your removable hard disk in “Disk Utility” software, create a new APFS container in APFS partition, name it “StarRail” (random name). Copy the Data folder from “Honkai: Star Rail” in the Containers folder to the StarRail container on your desktop, use “sudo umount -f /Volumes/StarRail”, then “sudo mount_apfs /dev/disk5s3 /Users/abc/Library/Containers/com.miHoYo.hkrpg” (com.miHoYo.hkrpg is the name and extension of the folder displaying the synopsis of Honkai: Star Rail under the Containers folder, and disk5s3 is the number of my StarRail container, which is different for everyone). Launch “Honkai: Star Rail” in PlayCover, if the system prompts the game to want to access the files on the external hard disk it means the move was successful, you can save space by deleting the Data inside the “Honkai: Star Rail” folder in the Containers folder after launching the removable hard disk.

<img width="1276" alt="截屏2024-10-10 07 59 08" src="https://github.com/user-attachments/assets/a0da03e9-2b3a-4e4f-828c-0c762349c493">

4.其他说明，每次插入移动硬盘都需要进行以上操作，大家可以编写一个.sh文件自动执行，具体教程可自行搜索。本方法对于游戏更新，覆盖安装等都可以，亲测「原神」更新正常。其他软件也可以通过这种方法移动来节省空间。对于新安装游戏，只要你的PlayCover在安装游戏时是在移动硬盘中运行的，那么安装的游戏也是在移动硬盘中的，但是Containers文件夹中的游戏文件也需要重新设置一遍。

> 4. Other instructions, every time you insert a removable hard disk need to perform the above operations, you can write a .sh file automatically executed, specific tutorials can search for themselves. This method can be used for game updates, overlay installations, etc., and the “Genshin Impact” update is normal. Other software can also be moved in this way to save space. For newly installed games, as long as your PlayCover is running in the removable hard disk when installing the game, the installed game is also in the removable hard disk, but the game files in the Containers folder need to be reset again as well.

**Comment by BlueSkyXN** [comment]
关闭SIP会导致iOS App无法使用，两头堵了啊

**Reply by BlueSkyXN** [reply]
> 关闭SIP会导致iOS App无法使用，两头堵了啊

试了下只能用子目录挂载绕开SIP

https://www.bilibili.com/video/BV1i4mKYiEtt/

**Comment by KarimAkra** [comment]
This has been working greatly for me even for non-PlayCover apps
But recently when i switched to MacOS 15.17.1 it stopped working...

The game instantly crashes with this:

```
-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Incident Identifier: C5FBEDAB-B9DE-4D62-A016-58C6E46D04D9
CrashReporter Key:   6D7AB77B-7264-20AA-9285-93B55539010A
Hardware Model:      Mac16,10
Process:             bh3global [44395]
Path:                /Volumes/VOLUME/*/bh3global.app/bh3global
Identifier:          com.miHoYo.bh3global
Version:             8.4.0 (741)
AppStoreTools:       16F7
AppVariant:          1:MacFamily20,1:18
Code Type:           ARM-64 (Native)
Role:                Default
Parent Process:      launchd [1]
Coalition:           com.miHoYo.bh3global [4118]

Date/Time:           2025-10-05 01:59:05.8588 +0300
Launch Time:         2025-10-05 01:59:05.7643 +0300
OS Version:          macOS 15.7.1 (24G231)
Release Type:        User
Report Version:      104

Exception Type:  EXC_BREAKPOINT (SIGTRAP)
Exception Codes: 0x0000000000000001, 0x00000001ae0970d0
Termination Reason: SIGNAL 5 Trace/BPT trap: 5
Terminating Process: exc handler [44395]

Triggered by Thread:  0

Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_secinit.dylib       	       0x1ae0970d0 _libsecinit_appsandbox.cold.6 + 60
1   libsystem_secinit.dylib       	       0x1ae0962b4 _libsecinit_appsandbox + 1968
2   libsystem_trace.dylib         	       0x19fb9a6cc _os_activity_initiate_impl + 64
3   libsystem_secinit.dylib       	       0x1ae095ab0 _libsecinit_initializer + 80
4   libSystem.B.dylib             	       0x1ae0ad32c libSystem_initializer + 280
5   dyld                          	       0x19facaefc invocation function for block in dyld4::Loader::findAndRunAllInitializers(dyld4::RuntimeState&) const + 444
6   dyld                          	       0x19fb0789c invocation function for block in dyld3::MachOAnalyzer::forEachInitializer(Diagnostics&, dyld3::MachOAnalyzer::VMAddrConverter const&, void (unsigned int) block_pointer, void const*) const + 324
7   dyld                          	       0x19fb275cc invocation function for block in mach_o::Header::forEachSection(void (mach_o::Header::SectionInfo const&, bool&) block_pointer) const + 240
8   dyld                          	       0x19fb24358 mach_o::Header::forEachLoadCommand(void (load_command const*, bool&) block_pointer) const + 208
9   dyld                          	       0x19fb25a98 mach_o::Header::forEachSection(void (mach_o::Header::SectionInfo const&, bool&) block_pointer) const + 124
10  dyld                          	       0x19fb0736c dyld3::MachOAnalyzer::forEachInitializer(Diagnostics&, dyld3::MachOAnalyzer::VMAddrConverter const&, void (unsigned int) block_pointer, void const*) const + 516
11  dyld                          	       0x19facacb4 dyld4::Loader::findAndRunAllInitializers(dyld4::RuntimeState&) const + 176
12  dyld                          	       0x19fad6530 dyld4::PrebuiltLoader::runInitializers(dyld4::RuntimeState&) const + 44
13  dyld                          	       0x19faec8f0 dyld4::APIs::runAllInitializersForMain() + 88
14  dyld                          	       0x19faafe00 dyld4::prepare(dyld4::APIs&, mach_o::Header const*) + 3092
15  dyld                          	       0x19faaf1d8 dyld4::start(dyld4::KernelArgs*, void*, void*)::$_0::operator()() const + 236
16  dyld                          	       0x19faaeb4c start + 6000


Thread 0 crashed with ARM Thread State (64-bit):
    x0: 0x00000000000003e9   x1: 0x0000000121008625   x2: 0xfffffffffffffff9   x3: 0x000000016f5efff9
    x4: 0xffffffffb1a1860c   x5: 0x0000000000000010   x6: 0x000000016f5efbf0   x7: 0x0000000000000000
    x8: 0x000000020ba90f50   x9: 0xa9accf3c6db700d7  x10: 0x00000000000007ff  x11: 0x00000000000003e9
   x12: 0x00000000000003e9  x13: 0x000000012100825c  x14: 0x9e3779b97f4a7c55  x15: 0x00000000ffffffff
   x16: 0x000000019fc4f674  x17: 0x000000020ee03058  x18: 0x0000000000000000  x19: 0x00000001210086b8
   x20: 0x000000016f5efc50  x21: 0x0000600000140050  x22: 0x0000600000140092  x23: 0x00000000000003e8
   x24: 0x00000001ae097f56  x25: 0x00000001ae097fa6  x26: 0x0000000000000000  x27: 0x0000000000000001
   x28: 0x00000001ae097fea   fp: 0x000000016f5efc10   lr: 0x00000001ae0970c4
    sp: 0x000000016f5efbf0   pc: 0x00000001ae0970d0 cpsr: 0x60000000
   far: 0x0000000000000000  esr: 0xf2000001 (Breakpoint) brk 1

Binary Images:
       0x100808000 -        0x10d37ffff com.miHoYo.bh3global (8.4.0) <56dce585-8c63-3f6c-b3d1-ac4969cf3ca3> /Volumes/VOLUME/*/bh3global.app/bh3global
       0x10f874000 -        0x10faaffff com.hoyoverse.sdk.Telemetry (2.0.14) <4c4c4452-5555-3144-a140-b7f7d84de361> /Volumes/VOLUME/*/bh3global.app/Frameworks/Telemetry.framework/Telemetry
       0x1ae095000 -        0x1ae0982c7 libsystem_secinit.dylib (*) <63965812-869c-3c3c-a950-fa563ce8eb1e> /usr/lib/system/libsystem_secinit.dylib
       0x19fb93000 -        0x19fbae7ff libsystem_trace.dylib (*) <9dbbee92-6c08-39f7-a05e-5054a79d28f6> /usr/lib/system/libsystem_trace.dylib
       0x1ae0ac000 -        0x1ae0adca3 libSystem.B.dylib (*) <068263f7-cf15-3e1c-a305-5df8f9c51ad5> /usr/lib/libSystem.B.dylib
       0x19faa8000 -        0x19fb4357b dyld (*) <8598f4c3-8cc9-38ac-b886-4d0fdf275a30> /usr/lib/dyld
               0x0 - 0xffffffffffffffff ??? (*) <00000000-0000-0000-0000-000000000000> ???
       0x19fc4e000 -        0x19fc9506f libsystem_malloc.dylib (*) <3ce1e59c-91c8-3259-93d8-7ded66842979> /usr/lib/system/libsystem_malloc.dylib

EOF
```
